pkg/preparer/orchestrate.go

```diff
@@ -222,83 +304,14 @@ func (p *Preparer) handlePods(podChan <-chan ManifestPair, quit <-chan struct{})
-				// pods that are in the blacklist for this preparer shall not use the
-				// preparer's log exec. Instead, they will use the default svlogd logexec.
-				for _, podID := range p.logBridgeBlacklist {
-					if pod.Id.String() == podID {
-						effectiveLogBridgeExec = svlogdExec
-						break
-					}
-				}
-				pod.SetLogBridgeExec(effectiveLogBridgeExec)
-				pod.SetFinishExec(p.finishExec)
-
-				// podChan is being fed values gathered from a consul.Watch() in
-				// WatchForPodManifestsForNode(). If the watch returns a new pair of
-				// intent/reality values before the previous change has finished
-				// processing in resolvePair(), the reality value will be stale. This
-				// leads to a bug where the preparer will appear to update a package
-				// and when that is finished, "update" it again.
-				//
-				// Example ordering of bad events:
-				// 1) update to /intent for pod A comes in, /reality is read and
-				// resolvePair() handles it
-				// 2) before resolvePair() finishes, another /intent update comes in,
-				// and /reality is read but hasn't been changed. This update cannot
-				// be processed until the previous resolvePair() call finishes, and
-				// updates /reality. Now the reality value used here is stale. We
-				// want to refresh our /reality read so we don't restart the pod if
-				// intent didn't change between updates.
-				//
-				// The correct solution probably involves watching reality and intent
-				// and feeding updated pairs to a control loop.
-				//
-				// This is a quick fix to ensure that the reality value being used is
-				// up-to-date. The de-bouncing logic in this method should ensure that the
-				// intent value is fresh (to the extent that Consul is timely). Fetching
-				// the reality value again ensures its freshness too.
-				if nextLaunch.PodUniqueKey == "" {
-					// legacy pod, get reality manifest from reality tree
-					reality, _, err := p.store.Pod(consul.REALITY_TREE, p.node, nextLaunch.ID)
-					if err == pods.NoCurrentManifest {
-						nextLaunch.Reality = nil
-					} else if err != nil {
-						manifestLogger.WithError(err).Errorln("Error getting reality manifest")
-						break
-					} else {
-						nextLaunch.Reality = reality
-					}
-				} else {
-					// uuid pod, get reality manifest from pod status
-					status, _, err := p.podStatusStore.Get(nextLaunch.PodUniqueKey)
-					switch {
-					case err != nil && !statusstore.IsNoStatus(err):
-						manifestLogger.WithError(err).Errorln("Error getting reality manifest from pod status")
-						break
-					case statusstore.IsNoStatus(err):
-						nextLaunch.Reality = nil
-					default:
-						manifest, err := manifest.FromBytes([]byte(status.Manifest))
-						if err != nil {
-							manifestLogger.WithError(err).Errorln("Error parsing reality manifest from pod status")
-							break
-						}
-						nextLaunch.Reality = manifest
-					}
+				ok, err := p.preparePod(nextLaunch, pod, manifestLogger)
```

Comment:
What is the advantage of two return values here? They seem to be used for the same purpose. `!ok` is never treated differently than `err != nil` so I don't see the purpose of two different failure cases.

Reply:
Because in some cases when there is no error, we also need to break the process.